src/SubscriptionStore.h

```diff
@@ -29,40 +33,194 @@ class SubscriptionStore
+        std::lock_guard<std::mutex> lock(subscriptionsStoreMutex);
+        bool success = true;
+        if (subscriptionsByConnection.count(connection) <= 0)
+        {
+            spdlog::error(
+                "Attempt to remove subscription for connection {} / channel {}, "
+                "but no subscriptions for this connection exist",
+                connection->GetHostname(),
+                channelId);
+            success = false;
+        }
+        else
+        {
+            auto& subs = subscriptionsByConnection[connection];
+            int numSubsErased = 0;
+            for (auto it = subs.begin(); it != subs.end();)
+            {
+                auto& sub = *it;
+                if (sub->ChannelId == channelId)
+                {
+                    it = subs.erase(it);
+                    ++numSubsErased;
+                }
+                else
+                {
+                    ++it;
+                }
+            }
+            if (numSubsErased == 0)
+            {
+                spdlog::error(
+                    "Attempt to remove non-existant subscription for connection {} to channel {}",
+                    connection->GetHostname(),
+                    channelId);
+                success = false;
+            }
+        }
+
+        if (subscriptionsByChannel.count(channelId) <= 0)
+        {
+            spdlog::error(
+                "Attempt to remove subscription for connection {} / channel {}, "
+                "but no subscriptions to this channel exist",
+                connection->GetHostname(),
+                channelId);
+            success = false;
+        }
+        else
+        {
+            auto& subs = subscriptionsByChannel[channelId];
+            int numSubsErased = 0;
+            for (auto it = subs.begin(); it != subs.end();)
+            {
+                auto& sub = *it;
+                if (sub->SubscribedConnection == connection)
+                {
+                    it = subs.erase(it);
+                    numSubsErased++;
+                }
+                else
+                {
+                    ++it;
+                }
+            }
+
+            if (numSubsErased == 0)
+            {
+                spdlog::error(
+                    "Attempt to remove non-existant subscription on channel {} for connection {}",
+                    connection->GetHostname(),
+                    channelId);
+                success = false;
+            }
+        }
+        return success;
+    }
 
     /**
      * @brief Get the list of channel subscriptions that exist for a connection
      * @param connection connection to fetch subscribed channels for
      * @return std::set<ftl_channel_id_t> set of channels this connection is subscribed to
      */
-    std::set<ftl_channel_id_t> GetSubscribedChannels(std::shared_ptr<IConnection> connection);
+    std::vector<ChannelSubscription<TConnection>> GetSubscriptions(std::shared_ptr<TConnection> connection)
```

Comment:
No need to change the code right now, but my style sense tingled for passing shared_ptrs everywhere. Functionally it's fine but C++ best practice is to only pass the shared pointer when you intended to "manipulate lifetimes" and to otherwise use references.

Their justification is more compatibility and partly performance, but I think more important is being clear about how this function uses the parameter. i.e. using a reference indicates access to the parameter will not escape the lifetime of this function while a shared_ptr parameter implies a new reference could be made that will outlive the function.

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-smartptrparam

Reply:
Yeah, I think this is bad practice that I've inherited from the ref pointer world of C++/CX. 

Sadly I think this anti-pattern is present throughout the project, so I think we'll need to sit and put some thought around lifetime management. At the least I have a unit test that makes sure that `IConnection` instances are properly destructed when the connection is closed, but we can do better.

Let's discuss this more in detail and lock down a convention for the project (and maybe open work items to track the debt)

Reply:
totally, CX is it's own beast.

👍 Good call, if you've been consistent so far best to stick with it, working code and tests are a thousand times more important imo than worrying about style.

Maybe I'll open an issue about discussion convention after launch (or switching to :crab: rust haha), but no need to add tracking overhead right now imo while the codebase is still small!

Reply:
I opened a tracking issue here: #10